### ISSUE TITLE
Panic policy

### ISSUE THREAD

**Issue by teozkr** [kube-rs/kube-rs#634]
Currently, we have a bunch of helpers that can implicitly panic the program when their invariants aren't upheld externally, such as [`ResourceExt::name`](https://docs.rs/kube/0.60.0/kube/trait.ResourceExt.html#tymethod.name). These essentially allow the apiserver (or any other source of `Resource`-`impl`ing objects) to panic the whole process.

Are these a good idea to encourage? Should we at least ban their transitive usage within the core `kube` crates?

If we do keep them around, we should probably at the very least document a policy for which invariants should be allowed to cause `kube` to panic. For example, it's worth considering (in escalating order of exposure):

1. private `kube` invariants
2. `kube` invariants exposed to the application (for example: any object to be created on the apiserver must have a non-empty `.metadata.name` xor `.metadata.generate_name`)
3. apiserver invariants (for example: any object returned by the apiserver must have a non-empty `.metadata.name`)
4. object-level expectations (for example: `Pod` will pretty much always have >=1 container)

**Comment by teozkr** [comment]
Going into personal opinion, I think 1 is pretty much unavoidable, 2 is often undesirable but may be the least bad option, and 3+ are walking and talking DoS vulnerabilities that we ought to discourage where possible.

**Comment by clux** [comment]
Yeah, that is a good assessment, and possibly the `ResourceExt::name` case has probably gone a bit far for convenience internally. We probably should not use it in the general case if it can bite users who use `generate_name`, although I am unsure if

How do you propose we document this policy? There was a crate that let us ban usage of certain methods from certain paths (but can't remember the name of it), that could possibly deal with an internal limit? 

There is also crates such as `pre` + `contracts`: https://docs.rs/pre/0.2.0/pre/ or https://docs.rs/contracts/0.6.2/contracts/
that allow us to put `#[pre]`-conditions that we have to `#[assure]` at the call-site. But not sure if that's super useful if it's exported.

**Comment by teozkr** [comment]
For documenting the policy itself, probably throw it in a document in the repo and maybe link to it from CONTRIBUTING?

For enforcing it, I'd probably go as far as deprecating the offending exceptions.

For guidance on what to do instead, one somewhat more ergonomic way would be to define a common helper struct with the fields you need, and a `TryFrom<ObjectMeta>`. That should let you centralize the error checking in one place.. There might be value in providing a derive macro to help with that process...

IIRC I've sometimes used `ObjectRef<K>` as that helper type, fwiw.

**Comment by teozkr** [comment]
Not sure I see how `ObjectReference` helps.. its fields are as `Option`al as `ObjectMeta`'s as far as I can tell?
